nexus/db-queries/src/db/datastore/inventory.rs

```diff
@@ -0,0 +1,914 @@
+/// details.
+#[must_use = "Queries must be executed"]
+struct InvCabooseInsert {
+    // fields used to look up baseboard id
+    baseboard_part_number: String,
+    baseboard_serial_number: String,
+
+    // fields used to look up caboose id
+    caboose_board: String,
+    caboose_git_commit: String,
+    caboose_name: String,
+    caboose_version: String,
+
+    // literal values for the rest of the inv_caboose columns
+    collection_id: Uuid,
+    time_collected: DateTime<Utc>,
+    source: String,
+    which: CabooseWhich,
+
+    // These are Diesel structures representing table names in the "from" or
+    // "into" parts of queries (e.g., "SELECT FROM tablename" or "INSERT INTO
+    // tablename").  We need this in `walk_ast()` below, but they must outlive
+    // `walk_ast()`, so they need to be created ahead of time.
+    //
+    // TODO-cleanup These Diesel-internal types are nasty.  It's not clear how
+    // else to do this.
+    from_hw_baseboard_id:
+        diesel::internal::table_macro::StaticQueryFragmentInstance<
+            db::schema::hw_baseboard_id::table,
+        >,
+    from_sw_caboose: diesel::internal::table_macro::StaticQueryFragmentInstance<
+        db::schema::sw_caboose::table,
+    >,
+    into_inv_caboose:
+        diesel::internal::table_macro::StaticQueryFragmentInstance<
+            db::schema::inv_caboose::table,
+        >,
+}
+
+impl InvCabooseInsert {
+    pub fn new(
+        collection_id: Uuid,
+        baseboard: &BaseboardId,
+        found_caboose: &CabooseFound,
+        which: CabooseWhich,
+    ) -> InvCabooseInsert {
+        InvCabooseInsert {
+            baseboard_part_number: baseboard.part_number.clone(),
+            baseboard_serial_number: baseboard.serial_number.clone(),
+            caboose_board: found_caboose.caboose.board.clone(),
+            caboose_git_commit: found_caboose.caboose.git_commit.clone(),
+            caboose_name: found_caboose.caboose.name.clone(),
+            caboose_version: found_caboose.caboose.version.clone(),
+            collection_id,
+            time_collected: found_caboose.time_collected,
+            source: found_caboose.source.clone(),
+            which,
+            from_hw_baseboard_id: db::schema::hw_baseboard_id::table
+                .from_clause(),
+            from_sw_caboose: db::schema::sw_caboose::table.from_clause(),
+            // It sounds a little goofy to use "from_clause()" when this is
+            // really part of an INSERT.  But really this just produces the
+            // table name as an identifier.  This is the same for both "FROM"
+            // and "INSERT" clauses.  And diesel internally does the same thing
+            // here (see the type of `InsertStatement::into_clause`).
+            into_inv_caboose: db::schema::inv_caboose::table.from_clause(),
+        }
+    }
+}
+
+impl diesel::query_builder::QueryFragment<diesel::pg::Pg> for InvCabooseInsert {
+    fn walk_ast<'b>(
```

Comment:
I have no useful suggestion here, but would like to register my complaint that this is much harder to read than the equivalent query written out in raw SQL would be.